drivers/serial/stm32-uart.c

```diff
@@ -0,0 +1,91 @@
+#include <sys/types.h>
+
+#include <kernel/cbuf.h>
+#include <kernel/fs.h>
+#include <kernel/irq.h>
+#include <kernel/serial.h>
+#include <kernel/types.h>
+
+#include <cmsis.h>
+#include "platform.h"
+
+static struct cbuf_info cbuf;
+static char buf[16];
+
+static int stm32_getc(struct serial_info *serial, char *c)
+{
+    serial->rx_count--;
+    cbuf_getc(&cbuf, c);
+
+    return 0;
+}
+
+static int stm32_putc(struct serial_info *serial, char c)
+{
+    USART_TypeDef *uart = serial->priv;
+
+    while (!(uart->SR & USART_SR_TXE))
+        ;
+    uart->DR = c;
+
+    return 0;
+}
+
+static int stm32_puts(struct serial_info *serial,
+                      size_t len,
+                      size_t *retlen,
+                      const char *buf)
+{
+    *retlen = len;
+    for (int i = 0; len > 0; len--, i++)
+        stm32_putc(serial, buf[i]);
+
+    return 0;
+}
+
+static struct serial_ops stm32_ops = {
+    .serial_getc = stm32_getc,
+    .serial_putc = stm32_putc,
+    .serial_puts = stm32_puts,
+};
+
+struct serial_info stm32_info = {
+    .priv = USARTx,
+    .ops = &stm32_ops,
+};
+
+static void stm32_uartx_isr(void)
+{
+    if (USARTx->SR & USART_SR_RXNE) {
+        char c = (char) USARTx->DR;
+        cbuf_putc(&cbuf, c);
+        stm32_info.rx_count++;
+        serial_activity_callback(&stm32_info);
+    }
+}
+
+extern const struct file_operations serialchar_fops;
+
+static struct inode stm32_inode = {
+    .i_fop = &serialchar_fops,
+    .i_private = &stm32_info,
+};
+
+int serial_init(void)
```

Comment:
It should be `stm32_serial_init` rather than the plain one.

Reply:
It should be, but `serial_init()` still needs to be implemented in `stm32-uart.c` for serial interface.

Reply:
No, `serial_init()` should never exist whatever in directory `drivers`. Instead, platform logic manipulates the relationship and coordinate.

Reply:
Code implemented in `drivers` is expected to expose symbols for platforms.